File: iosApp/Services/ExportService.swift
import Foundation
import Observation
import UIKit
@preconcurrency import Shared

/// Failures raised by [ExportService]. All paths emit an `export_failed`
/// audit entry before propagating.
enum ExportError: LocalizedError {
    case clinicianMissing
    case clinicConfigMissing
    case zipFailed(reason: String)
    case writeFailed(reason: String)

    var errorDescription: String? {
        switch self {
        case .clinicianMissing:
            "No clinician profile found. Complete onboarding before exporting."
        case .clinicConfigMissing:
            "Clinic configuration not found in the audit log. Complete admin provisioning first."
        case .zipFailed(let reason):
            "Could not create the export archive: \(reason)"
        case .writeFailed(let reason):
            "Could not write the export bundle: \(reason)"
        }
    }
}

/// Phase 13 export-bundle service (spec §14). The clinical-correctness logic
/// (JSON canonicalisation + HMAC signing) lives in the shared Kotlin module;
/// this `@Observable @MainActor` class is the iOS-side adapter that:
///
///   1. Fetches typed inputs from the four repositories + `SettingsStore`.
///   2. Converts SwiftData entities to the shared `Exported*` DTOs
///      (timestamps go through `ISO8601DateFormatter` with the
///      `.withInternetDateTime` options so the format matches the Android
///      `Instant.toString()` form byte-for-byte).
///   3. Calls `ExportBundleBuilder.build(...)` to produce the canonical
///      signed JSON.
///   4. Packs the JSON + a short `README.txt` into a ZIP via a tiny stored-
///      mode writer (`MinimalZipWriter`) and returns the file URL in a temp
///      directory.
///   5. Emits the `export_initiated` / `export_completed` / `export_failed`
///      audit chain per spec §14.
@Observable
@MainActor
final class ExportService {

    private let predictions: PredictionRepository
    private let audits: AuditRepository
    private let clinicians: ClinicianRepository
    private let consents: ConsentRepository
    private let settings: SettingsStore
    private let auditLog: AuditLog

    /// `UserDefaults` key for the cached device UUID fallback (spec §14
    /// wants a stable device identifier; `identifierForVendor` resets on
    /// app uninstall + reinstall which matches device-reset semantics, but
    /// in unit tests there is no UIDevice attached so we cache the value).
    private let deviceUuidDefaultsKey = "com.malaria.detector.exportDeviceUuid"

    init(
        predictions: PredictionRepository,
        audits: AuditRepository,
        clinicians: ClinicianRepository,
        consents: ConsentRepository,
        settings: SettingsStore,
        auditLog: AuditLog
    ) {
        self.predictions = predictions
        self.audits = audits
        self.clinicians = clinicians
        self.consents = consents
        self.settings = settings
        self.auditLog = auditLog
    }

    /// Produces a signed ZIP bundle and returns its file URL. The caller
    /// presents the share sheet (`UIActivityViewController`) over the URL.
    ///
    /// Side effects:
    ///   - On entry: writes an `export_initiated` audit entry.
    ///   - On success: writes `export_completed` with `size` and `signature`
    ///     metadata.
    ///   - On any thrown error: writes `export_failed` with `reason`
    ///     metadata.
    func generateBundle() async throws -> URL {
        let actor = try? clinicians.current()
        let actorId = actor?.actorId ?? "unknown"
        let actorRole = actor?.role ?? "unknown"

        auditLog.write(.exportInitiated, actorId: actorId, actorRoleAtTime: actorRole)

        do {
            let build = try await buildBundle(actorId: actorId)
            try MinimalZipWriter.writeArchive(
                entries: [
                    (name: "export.json", data: Data(build.signedJson.utf8)),
                    (name: "README.txt", data: Data(build.readme.utf8))
                ],
                to: build.bundleURL
            )
            let attrs = try? FileManager.default.attributesOfItem(atPath: build.bundleURL.path)
            let size = (attrs?[.size] as? NSNumber)?.intValue ?? 0
            auditLog.write(
                .exportCompleted,
                actorId: actorId,
                actorRoleAtTime: actorRole,
                metadata: ["size": String(size), "signature": build.signature]
            )
            return build.bundleURL
        } catch {
            auditLog.write(
                .exportFailed,
                actorId: actorId,
                actorRoleAtTime: actorRole,
                metadata: ["reason": error.localizedDescription]
            )
            throw error
        }
    }

    // MARK: - Internals

    /// Output of the JSON+naming step. The ZIP write is a separate step so
    /// the audit-completion metadata can record the file size and signature
    /// in one place.
    struct BundleBuild {
        let bundleURL: URL
        let signedJson: String
        let signature: String
        let readme: String
    }

    private func buildBundle(actorId: String) async throws -> BundleBuild {
        guard let clinicName = settings.clinicName,
              let jurisdiction = settings.jurisdiction,
              let lawfulBasis = settings.lawfulBasis else {
            throw ExportError.clinicConfigMissing
        }

        let exportDate = Date()
        let exportTimestamp = Self.iso8601(exportDate)

        // ----- Snapshot persistence into the shared DTO types -----------
        let predictionsRows = try predictions.recent(limit: Int.max)
        let auditRows = try audits.recent(limit: Int.max)
        let allProfiles = try [clinicians.current()].compactMap { $0 }
        let consentRecords = try consents.records(for: actorId)

        let dtoPredictions = predictionsRows.map(Self.toDTO(_:))
        let dtoAudits = auditRows.map(Self.toDTO(_:))
        let dtoProfiles = allProfiles.map(Self.toDTO(_:))

        let summary = ExportSummary(
            predictionCount: Int32(predictionsRows.count),
            sessionCount: Int32(Set(predictionsRows.map { $0.sessionId }).count),
            auditEntryCount: Int32(auditRows.count),
            consentRecordCount: Int32(consentRecords.count),
            firstPredictionAt: predictionsRows.map { Self.iso8601($0.timestamp) }.min(),
            lastPredictionAt: predictionsRows.map { Self.iso8601($0.timestamp) }.max()
        )

        // ----- Sign + serialize via the shared module -------------------
        let deviceUuid = Self.deviceUuid(defaultsKey: deviceUuidDefaultsKey)
        let signedJson = ExportBundleBuilder().build(
            exportTimestamp: exportTimestamp,
            exportedByActorId: actorId,
            deviceUuid: deviceUuid,
            platform: "ios",
            clinicName: clinicName,
            jurisdiction: jurisdiction,
            lawfulBasis: lawfulBasis,
            appVersion: BuildEnvironment.appVersion,
            osVersion: BuildEnvironment.osVersion,
            summary: summary,
            clinicianProfiles: dtoProfiles,
            predictions: dtoPredictions,
            auditLog: dtoAudits
        )

        let signature = Self.extractSignature(from: signedJson) ?? "unknown"

        let timestampSlug = Self.timestampSlug(exportDate)
        let prefix = String(deviceUuid.prefix(8))
        let filename = "malaria-detector-export-\(prefix)-\(timestampSlug).zip"
        let bundleURL = FileManager.default.temporaryDirectory
            .appendingPathComponent(filename)
        try? FileManager.default.removeItem(at: bundleURL)

        let readme = Self.readmeText(
            exportTimestamp: exportTimestamp,
            clinicName: clinicName,
            summary: summary
        )

        return BundleBuild(
            bundleURL: bundleURL,
            signedJson: signedJson,
            signature: signature,
            readme: readme
        )
    }

    private static func extractSignature(from json: String) -> String? {
        let marker = "\"signature\":\""
        guard let start = json.range(of: marker, options: .backwards) else { return nil }
        let after = json[start.upperBound...]
        guard let end = after.firstIndex(of: "\"") else { return nil }
        return String(after[..<end])
    }

    // MARK: - DTO conversion

    private static func toDTO(_ p: Prediction) -> ExportedPrediction {
        ExportedPrediction(
            id: p.id,
            sessionId: p.sessionId,
            timestamp: iso8601(p.timestamp),
            modelId: p.modelId,
            modelVersion: p.modelVersion,
            parasitizedProb: p.parasitizedProb,
            uninfectedProb: p.uninfectedProb,
            label: p.label,
            threshold: p.threshold,
            flaggedForReview: p.flaggedForReview,
            inferenceMs: Int32(p.inferenceMs),
            imageHash: p.imageHash,
            clinicianOverride: p.clinicianOverride,
            overrideContext: p.overrideContext,
            duplicateOfId: p.duplicateOfId,
            sessionLabel: p.sessionLabel,
            appVersion: p.appVersion,
            osVersion: p.osVersion
        )
    }

    private static func toDTO(_ e: AuditEntry) -> ExportedAuditEntry {
        ExportedAuditEntry(
            id: e.id,
            seq: Int64(e.seq),
            timestamp: iso8601(e.timestamp),
            actorId: e.actorId,
            actorRoleAtTime: e.actorRoleAtTime,
            action: e.action,
            resourceType: e.resourceType,
            resourceId: e.resourceId,
            metadataJson: e.metadataJson,
            overrideContext: e.overrideContext,
            overrideReason: e.overrideReason,
            overrideNotes: e.overrideNotes,
            contextReviewed: e.contextReviewed.map { KotlinBoolean(value: $0) },
            overrideActorInitials: e.overrideActorInitials,
            appVersion: e.appVersion,
            osVersion: e.osVersion
        )
    }

    private static func toDTO(_ c: ClinicianProfile) -> ExportedClinicianProfile {
        ExportedClinicianProfile(
            actorId: c.actorId,
            role: c.role,
            initials: c.initials,
            enrolledAt: iso8601(c.enrolledAt),
            biometricEnrolled: c.biometricEnrolled
        )
    }

    // MARK: - Formatters

    private static let iso8601Formatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    private static func iso8601(_ date: Date) -> String {
        iso8601Formatter.string(from: date)
    }

    private static func timestampSlug(_ date: Date) -> String {
        // `2026-05-21T14-32-08Z` form — derived from the ISO-8601 with the
        // colons replaced (filesystem-safe).
        let raw = iso8601(date)
        return raw.replacingOccurrences(of: ":", with: "-")
    }

    static func deviceUuid(defaultsKey: String) -> String {
        if let vendor = UIDevice.current.identifierForVendor?.uuidString {
            return vendor
        }
        // Test or simulator fallback — cache a UUID in UserDefaults so the
        // export is reproducible across runs.
        let defaults = UserDefaults.standard
        if let cached = defaults.string(forKey: defaultsKey), !cached.isEmpty {
            return cached
        }
        let generated = UUID().uuidString
        defaults.set(generated, forKey: defaultsKey)
        return generated
    }

    private static func readmeText(
        exportTimestamp: String,
        clinicName: String,
        summary: ExportSummary
    ) -> String {
        """
        Malaria Detector export bundle
        ==============================
        Exported at:  \(exportTimestamp)
        Clinic:       \(clinicName)
        Predictions:  \(summary.predictionCount)
        Sessions:     \(summary.sessionCount)
        Audit rows:   \(summary.auditEntryCount)
        Consents:     \(summary.consentRecordCount)

        Contents:
          - export.json   — signed bundle (HMAC-SHA256, see spec §14)
          - README.txt    — this file

        Verifying the signature:
          1. Open export.json, strip the trailing `,"signature":"<hex>"}`.
          2. Re-serialise via the shared ExportBundleBuilder JSON config.
          3. Recompute the HMAC over the unsigned form using the device
             UUID + exportTimestamp salt.
          4. Compare to the original signature field.

        Generated by Malaria Detector v1. No images are included in this
        bundle — imageHash is the only durable trace of the analysed
        cells (spec §8).
        """
    }
}

File: androidApp/src/main/kotlin/com/malaria/android/services/ExportService.kt
package com.malaria.android.services

import android.content.ContentResolver
import android.content.Context
import android.provider.Settings
import com.malaria.android.data.BuildEnvironment
import com.malaria.android.data.dao.AuditDao
import com.malaria.android.data.dao.ClinicianDao
import com.malaria.android.data.dao.ConsentDao
import com.malaria.android.data.dao.PredictionDao
import com.malaria.android.data.entities.AuditAction
import com.malaria.android.data.entities.AuditEntry
import com.malaria.android.data.entities.ClinicianProfile
import com.malaria.android.data.entities.Prediction
import com.malaria.export.ExportBundleBuilder
import com.malaria.export.ExportPlatform
import com.malaria.export.ExportSummary
import com.malaria.export.ExportedAuditEntry
import com.malaria.export.ExportedClinicianProfile
import com.malaria.export.ExportedPrediction
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.datetime.Clock
import kotlinx.datetime.Instant
import java.io.File
import java.util.UUID
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream

/**
 * Phase 13 export-bundle service (spec §14). Mirrors
 * `iosApp/Services/ExportService.swift` step-for-step. The clinical-
 * correctness logic — JSON canonicalisation, HMAC signing, stable field
 * order — lives in the shared Kotlin module's
 * [com.malaria.export.ExportBundleBuilder]; this class is the Android-side
 * adapter that:
 *
 *   1. Loads typed inputs from the four DAOs + [SettingsStore].
 *   2. Converts Room entities to the shared `Exported*` DTOs (timestamps go
 *      through `Instant.toString()` so the format matches the iOS
 *      `ISO8601DateFormatter` form byte-for-byte).
 *   3. Calls `ExportBundleBuilder.build(...)` for the canonical signed JSON.
 *   4. Packs the JSON + a short `README.txt` into a ZIP via
 *      [java.util.zip.ZipOutputStream] and saves it under `context.cacheDir`.
 *   5. Emits the `export_initiated` / `export_completed` / `export_failed`
 *      audit chain per spec §14.
 *
 * The audit chain is the only side effect besides the temp-file write. The
 * caller (the DataManagementView composable) shares the resulting file via
 * `Intent.ACTION_SEND` + a `FileProvider`-issued content `Uri`.
 */
class ExportService(
    private val predictions: PredictionDao,
    private val audits: AuditDao,
    private val clinicians: ClinicianDao,
    private val consents: ConsentDao,
    private val settings: SettingsStore,
    private val auditLog: AuditLog,
    private val cacheRoot: File,
    private val clock: () -> Instant = { Clock.System.now() },
    private val deviceUuidProvider: () -> String,
) {

    /**
     * Production constructor wiring the device-scoped `Settings.Secure.ANDROID_ID`
     * UUID and the platform `Context.cacheDir`. Tests use the primary
     * constructor directly so neither dependency needs to be mocked.
     */
    constructor(
        context: Context,
        predictions: PredictionDao,
        audits: AuditDao,
        clinicians: ClinicianDao,
        consents: ConsentDao,
        settings: SettingsStore,
        auditLog: AuditLog,
    ) : this(
        predictions = predictions,
        audits = audits,
        clinicians = clinicians,
        consents = consents,
        settings = settings,
        auditLog = auditLog,
        cacheRoot = context.cacheDir,
        clock = { Clock.System.now() },
        deviceUuidProvider = { deriveDeviceUuid(context.contentResolver) },
    )

    private val _lastError = MutableStateFlow<Throwable?>(null)
    val lastError: StateFlow<Throwable?> = _lastError.asStateFlow()

    /** Produces a signed ZIP bundle and returns its [File]. */
    suspend fun generateBundle(): File {
        val actor = runCatching { clinicians.current() }.getOrNull()
        val actorId = actor?.actorId ?: "unknown"
        val actorRole = actor?.role ?: "unknown"

        auditLog.write(
            action = AuditAction.ExportInitiated,
            actorId = actorId,
            actorRoleAtTime = actorRole,
        )

        return try {
            val build = buildBundle(actorId = actorId)
            val zipFile = writeZip(build)
            val size = zipFile.length()
            auditLog.write(
                action = AuditAction.ExportCompleted,
                actorId = actorId,
                actorRoleAtTime = actorRole,
                metadata = mapOf(
                    "size" to size.toString(),
                    "signature" to build.signature,
                ),
            )
            _lastError.value = null
            zipFile
        } catch (t: Throwable) {
            _lastError.value = t
            auditLog.write(
                action = AuditAction.ExportFailed,
                actorId = actorId,
                actorRoleAtTime = actorRole,
                metadata = mapOf("reason" to (t.message ?: "unknown")),
            )
            throw t
        }
    }

    // -- Internals -------------------------------------------------------

    private data class BundleBuild(
        val signedJson: String,
        val readme: String,
        val signature: String,
        val filename: String,
    )

    private suspend fun buildBundle(actorId: String): BundleBuild {
        val clinicName = settings.clinicName.value
            ?: throw ExportException("Clinic configuration not found. Complete admin provisioning first.")
        val jurisdiction = settings.jurisdiction.value
            ?: throw ExportException("Jurisdiction not configured.")
        val lawfulBasis = settings.lawfulBasis.value
            ?: throw ExportException("Lawful basis not configured.")

        val now = clock()
        val exportTimestamp = now.toString()

        val predictionRows = predictions.recent(limit = Int.MAX_VALUE)
        val auditRows = audits.recent(limit = Int.MAX_VALUE)
        val clinicianRow = runCatching { clinicians.current() }.getOrNull()
        val consentRecords = if (clinicianRow != null) {
            runCatching { consents.records(clinicianRow.actorId) }.getOrDefault(emptyList())
        } else {
            emptyList()
        }

        val dtoPredictions = predictionRows.map { it.toDto() }
        val dtoAudits = auditRows.map { it.toDto() }
        val dtoProfiles = listOfNotNull(clinicianRow).map { it.toDto() }

        val summary = ExportSummary(
            predictionCount = predictionRows.size,
            sessionCount = predictionRows.map { it.sessionId }.toSet().size,
            auditEntryCount = auditRows.size,
            consentRecordCount = consentRecords.size,
            firstPredictionAt = predictionRows.minOfOrNull { it.timestamp }?.toString(),
            lastPredictionAt = predictionRows.maxOfOrNull { it.timestamp }?.toString(),
        )

        val deviceUuid = deviceUuidProvider()
        val signedJson = ExportBundleBuilder().build(
            exportTimestamp = exportTimestamp,
            exportedByActorId = actorId,
            deviceUuid = deviceUuid,
            platform = ExportPlatform.ANDROID,
            clinicName = clinicName,
            jurisdiction = jurisdiction,
            lawfulBasis = lawfulBasis,
            appVersion = BuildEnvironment.appVersion,
            osVersion = BuildEnvironment.osVersion,
            summary = summary,
            clinicianProfiles = dtoProfiles,
            predictions = dtoPredictions,
            auditLog = dtoAudits,
        )

        val signature = extractSignature(signedJson) ?: "unknown"
        val prefix = deviceUuid.take(8)
        val timestampSlug = exportTimestamp.replace(':', '-')
        val filename = "malaria-detector-export-$prefix-$timestampSlug.zip"

        val readme = buildReadme(
            exportTimestamp = exportTimestamp,
            clinicName = clinicName,
            summary = summary,
        )

        return BundleBuild(
            signedJson = signedJson,
            readme = readme,
            signature = signature,
            filename = filename,
        )
    }

    private fun writeZip(build: BundleBuild): File {
        val outputDir = File(cacheRoot, EXPORT_CACHE_SUBDIR).apply { mkdirs() }
        val zipFile = File(outputDir, build.filename)
        // Overwrite any stale file at the same name.
        if (zipFile.exists()) zipFile.delete()

        ZipOutputStream(zipFile.outputStream().buffered()).use { zip ->
            zip.putNextEntry(ZipEntry("export.json"))
            zip.write(build.signedJson.toByteArray(Charsets.UTF_8))
            zip.closeEntry()
            zip.putNextEntry(ZipEntry("README.txt"))
            zip.write(build.readme.toByteArray(Charsets.UTF_8))
            zip.closeEntry()
        }
        return zipFile
    }

    private fun extractSignature(json: String): String? {
        val marker = "\"signature\":\""
        val start = json.lastIndexOf(marker)
        if (start < 0) return null
        val from = start + marker.length
        val end = json.indexOf('"', from)
        if (end < 0) return null
        return json.substring(from, end)
    }

    private fun buildReadme(
        exportTimestamp: String,
        clinicName: String,
        summary: ExportSummary,
    ): String = """
        Malaria Detector export bundle
        ==============================
        Exported at:  $exportTimestamp
        Clinic:       $clinicName
        Predictions:  ${summary.predictionCount}
        Sessions:     ${summary.sessionCount}
        Audit rows:   ${summary.auditEntryCount}
        Consents:     ${summary.consentRecordCount}

        Contents:
          - export.json   — signed bundle (HMAC-SHA256, see spec §14)
          - README.txt    — this file

        Verifying the signature:
          1. Open export.json, strip the trailing `,"signature":"<hex>"}`.
          2. Re-serialise via the shared ExportBundleBuilder JSON config.
          3. Recompute the HMAC over the unsigned form using the device
             UUID + exportTimestamp salt.
          4. Compare to the original signature field.

        Generated by Malaria Detector v1. No images are included in this
        bundle — imageHash is the only durable trace of the analysed
        cells (spec §8).
    """.trimIndent()

    companion object {
        /** Subdirectory under `context.cacheDir` where bundles are stored. */
        const val EXPORT_CACHE_SUBDIR: String = "exports"

        /**
         * Derive the device UUID for the export bundle.
         *
         * `Settings.Secure.ANDROID_ID` is a 64-bit value scoped to the
         * signing key + user + device combination. It resets on factory
         * reset, which matches the spec §14 "device-reset semantics"
         * expectation. If the value is missing (rare; some emulators
         * return null), a fresh random UUID is generated — the bundle is
         * still verifiable; the only thing lost is per-device stability.
         */
        fun deriveDeviceUuid(resolver: ContentResolver): String {
            val androidId = runCatching {
                @Suppress("HardwareIds")
                Settings.Secure.getString(resolver, Settings.Secure.ANDROID_ID)
            }.getOrNull()
            return if (!androidId.isNullOrBlank()) androidId else UUID.randomUUID().toString()
        }
    }
}

/** Thrown when the bundle cannot be assembled (e.g. clinic not configured). */
class ExportException(message: String) : RuntimeException(message)

// -- DTO conversion -------------------------------------------------------

private fun Prediction.toDto(): ExportedPrediction = ExportedPrediction(
    id = id,
    sessionId = sessionId,
    timestamp = timestamp.toString(),
    modelId = modelId,
    modelVersion = modelVersion,
    parasitizedProb = parasitizedProb,
    uninfectedProb = uninfectedProb,
    label = label,
    threshold = threshold,
    flaggedForReview = flaggedForReview,
    inferenceMs = inferenceMs,
    imageHash = imageHash,
    clinicianOverride = clinicianOverride,
    overrideContext = overrideContext,
    duplicateOfId = duplicateOfId,
    sessionLabel = sessionLabel,
    appVersion = appVersion,
    osVersion = osVersion,
)

private fun AuditEntry.toDto(): ExportedAuditEntry = ExportedAuditEntry(
    id = id,
    seq = seq,
    timestamp = timestamp.toString(),
    actorId = actorId,
    actorRoleAtTime = actorRoleAtTime,
    action = action,
    resourceType = resourceType,
    resourceId = resourceId,
    metadataJson = metadataJson,
    overrideContext = overrideContext,
    overrideReason = overrideReason,
    overrideNotes = overrideNotes,
    contextReviewed = contextReviewed,
    overrideActorInitials = overrideActorInitials,
    appVersion = appVersion,
    osVersion = osVersion,
)

private fun ClinicianProfile.toDto(): ExportedClinicianProfile = ExportedClinicianProfile(
    actorId = actorId,
    role = role,
    initials = initials,
    enrolledAt = enrolledAt.toString(),
    biometricEnrolled = biometricEnrolled,
)
